url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/MCS.java

```diff
@@ -0,0 +1,177 @@
+package clarion.system;
+
+import java.util.HashSet;
+import java.util.HashMap;
+import java.util.Collection;
+
+/**
+ * This class implements the Meta Cognitive Subsystem (MCS) within CLARION. 
+ * It extends the AbstractSubsystem class.
+ * <p>
+ * <b>Usage:</b>
+ * <p>
+ * This class contains the modules of the MCS along with methods for getting and setting those modules.
+ * The user can decide, as desired, which modules to attach to the MCS and the CLARION Library will
+ * use those modules if they are attached, otherwise it wont. In this sense, the CLARION Library
+ * is dynamic and flexible about which modules appear within the MCS.
+ * <p>
+ * The modules of the MCS are accessed most commonly by the CLARION class through the perceive and act methods.
+ * Note that if a developer wished to write a currently unimplemented module into the MCS, the only ways
+ * to do so would be to either edit the CLARION library (not recommended) or write their own
+ * MCS and CLARION classes by extending those classes and then adding/overriding (respectively) the new 
+ * modules/perceive and act functions (again respectively).
+ * <p>
+ * Currently, the following modules have been implemented, and can be attached, as needed, within the MCS:
+ * <ul>
+ * <li>The ACS Level Probability Setting Module</li>
+ * <li>The Goal Selection Module </li>
+ * <li>The Judgment Correction Module </li>
+ * </ul>
+ * <p>
+ * Note that it is not required that a CLARION agent have an MCS at all. If the task being performed
+ * by the agent does not require an MCS, the user can choose simply not to attach the MCS to the
+ * CLARION agent and that agent will be able to operate quite successfully without this subsystem being
+ * specified (for some tasks).
+ * @version 6.0.6
+ * @author Nick Wilson
+ */
+public class MCS extends AbstractSubsystem{
+	
+	private HashMap <Class<? extends InterfaceMCSRunTimes>, HashSet <AbstractMetaCognitiveModule>> Modules = 
+		new HashMap<Class<? extends InterfaceMCSRunTimes>, HashSet <AbstractMetaCognitiveModule>> ();
+	
+	/**
+	 * Minimally initializes the MCS. This constructor builds a "frame" for this instance of
+	 * the MS from which all desired components can be attached. During initialization this
+	 * instance of the MCS will attach itself to the CLARION agent you specify.
+	 * @param Agent The agent to which the MCS is being attached.
+	 */
+	public MCS (CLARION Agent) 
+	{
+		super(Agent);
+	}
+	
+	/**
+	 * Adds the specified module to the MCS.
+	 * @param module The module to add.
+	 */
+	public void addModule (AbstractMetaCognitiveModule module)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 62-113 into a separate method.